include/extractor/restriction.hpp

```diff
@@ -5,111 +5,267 @@
+        return boost::get<InputNodeRestriction>(node_or_way);
     }
+
+    const InputNodeRestriction &asNodeRestriction() const
+    {
+        BOOST_ASSERT(!flags.way_type);
+        return boost::get<InputNodeRestriction>(node_or_way);
+    }
+};
+struct InputConditionalTurnRestriction : InputTurnRestriction
+{
+    std::vector<util::OpeningHours> condition;
 };
 
-struct CmpRestrictionContainerByFrom
+// OSRM manages restrictions based on node IDs which refer to the last node along the edge. Note
+// that this has the side-effect of not allowing parallel edges!
+struct NodeRestriction
 {
-    using value_type = InputRestrictionContainer;
-    bool operator()(const InputRestrictionContainer &a, const InputRestrictionContainer &b) const
+    NodeID from;
+    NodeID via;
+    NodeID to;
+
+    // check if all parts of the restriction reference an actual node
+    bool valid() const
+    {
+        return from != SPECIAL_NODEID && to != SPECIAL_NODEID && via != SPECIAL_NODEID;
+    };
+
+    std::string toString() const
     {
-        return a.restriction.from.way < b.restriction.from.way;
+        return "From " + std::to_string(from) + " via " + std::to_string(via) + " to " +
+               std::to_string(to);
     }
-    value_type max_value() const { return InputRestrictionContainer::max_value(); }
-    value_type min_value() const { return InputRestrictionContainer::min_value(); }
 };
 
-struct CmpRestrictionContainerByTo
+// A way restriction in the context of OSRM requires translation into NodeIDs. This is due to the
+// compression happening in the graph creation process which would make it difficult to track
+// way-ids over a series of operations. Having access to the nodes directly allows look-up of the
+// edges in the processed structures
+struct WayRestriction
+{
+    // a way restriction in OSRM is essentially a dual node turn restriction;
+    //
+    // |     |
+    // c -x- b
+    // |     |
+    // d     a
+    //
+    // from ab via bxc to cd: no_uturn
+    //
+    // Technically, we would need only a,b,c,d to describe the full turn in terms of nodes. When
+    // parsing the relation, though, we do not know about the final representation in the node-based
+    // graph for the restriction. In case of a traffic light, for example, we might end up with bxc
+    // not being compressed to bc. For that reason, we need to maintain two node restrictions in
+    // case a way restrction is not fully collapsed
+    NodeRestriction in_restriction;
+    NodeRestriction out_restriction;
+};
+
+// Wrapper for turn restrictions that gives more information on its type / handles the switch
+// between node/way/multi-way restrictions
+struct TurnRestriction
 {
-    using value_type = InputRestrictionContainer;
-    bool operator()(const InputRestrictionContainer &a, const InputRestrictionContainer &b) const
+    boost::variant<NodeRestriction, WayRestriction> node_or_way;
+    restriction_details::Bits flags;
```

Comment:
This is an open change I have to do. I was originally using the normal structs that were in the data and only then switched to a true variant to avoid reinterpreting data over multiple types. Need to update that.